[libc++] Implement P1885R12: `<text_encoding>`

smallp-o-p opened the pull request:
Resolves #105373, resolves #118371 and resolves #105332

- Implements `<text_encoding>`
- Adds availability macros for LLVM 23.
- The data is stored in three tables:
  - One giant string split by null-terminators to represent the aliases
  - An index table which stores indexes into the string, each entry representing the first character of an alias
  - Text encoding data, which stores an index to the index table, the MIB, and the number of aliases the encoding has.

Storing it in the above manner allows us to make significant savings in binary file size and required runtime storage for the data.

As required by the LLVM Project's AI use policy:
-  The implementation for `__get_locale_encoding(const char*)` for Windows has been developed with the assistance of AI.

smallp-o-p committed:
Implement P1885R12

frederick-vs-ja commented:
Thanks! I've edited the PR description to associate this PR to both issues.

cor3ntin commented on libcxx/src/text_encoding.cpp:
```diff
@@ -0,0 +1,49 @@
+  auto __make_locale = [](const char* __name) {
+    text_encoding __enc{};
+    if (auto __loc = __locale::__newlocale(LC_CTYPE_MASK, __name, static_cast<locale_t>(0))) {
+      if (const char* __codeset = nl_langinfo_l(CODESET, __loc)) {
```
You might want to check that you are in a POSIX environment here. `nl_langinfo_l` is not going to be available on windows, for example

cor3ntin commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+  }
+  _LIBCPP_HIDE_FROM_ABI constexpr text_encoding(id __i) _NOEXCEPT : __encoding_rep_(__find_encoding_data_by_id(__i)) {
+    if (__encoding_rep_->__name[0] != '\0')
+      std::copy_n(__encoding_rep_->__name, std::char_traits<char>::length(__encoding_rep_->__name), __name_);
```
I would just use strncpy here - but I don;t know what libc++ folks prefer

cor3ntin commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+
+      _LIBCPP_HIDE_FROM_ABI constexpr __iterator() noexcept = default;
+
+      _LIBCPP_HIDE_FROM_ABI constexpr value_type operator*() const {
+        if (__can_dereference())
+          return __data_->__name;
```
Maybe use an assert/precondition check here instead?

cor3ntin commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+      // Check if going past the encoding data list array and if the new index has the same id, if not then
+      // replace it with a sentinel "out-of-bounds" iterator.
+      _LIBCPP_HIDE_FROM_ABI constexpr __iterator& operator+=(difference_type __n) {
+        if (__data_) [[__likely__]] {
```
[[__likely__]]  likely doesn't do anything :)

cor3ntin commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+
+  [[nodiscard]] static text_encoding environment();
+
+  template <id __i>
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI static bool environment_is() {
+    return environment() == __i;
+  }
```
That function intended that it could be optimized - eg for utf8 - such that it would not access / odr-use the data table. But that implementation is fine, especially as a first pass

cor3ntin commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+  }
+
+  _LIBCPP_HIDE_FROM_ABI static constexpr __encoding_data __text_encoding_data[] = {
+      {1, ""},
+      {2, ""},
```
Are these useful?

cor3ntin reviewed (commented):
Exciting to see progress on this

I'm not a library maintainer, so take my comments for what they are worth :)


What's the Windows support for libc++ these days? environment is Posix specific at the moment

smallp-o-p commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+#  pragma GCC system_header
+#endif
+
+#if _LIBCPP_HAS_LOCALIZATION
```
I would think a user who is building libc++ without localization wouldn't have much use for a text encoding database, hence guarding it behind `_LIBCPP_HAS_LOCALIZATION`, however I'm not an expert. In the standard, localization and text_encoding are different subcategories of the text processing library, so it's more likely that the guard should be removed. Perhaps @cor3ntin could weigh in?

cor3ntin replied:
locale and encoding are unrelated, for the most part. so I would not disable the feature if locale is not available.  
*maybe* environment/environment_is would be considered to require locale?

smallp-o-p commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+    reserved                = 3000
+  };
+
+  using enum id;
```
Good catch. The first implementation of this enum (which I shamelessly copied) by @cor3ntin had `reserved`, and interestingly enough has two extra members: `NATSDANO = 33`, NATSDANOADD = 34` which are not present in the standard draft I have. @cor3ntin is this correct?

cor3ntin replied:
reserved is gone if I recall correctly.

the two ids you mentioned are called out explicitly 
https://eel.is/c++draft/text.encoding#general-2 (the reason being that in addition to being extremely desuet,  they have aliases that would have compared equal under comp_name)

smallp-o-p commented on libcxx/src/text_encoding.cpp:
```diff
@@ -0,0 +1,49 @@
+#  include <langinfo.h>
+#endif
+
+#if _LIBCPP_STD_VER >= 26
```
Pardon my unfamiliarity with the codebase, but wouldn't this prevent the library from being built, since it's built with `-std=c++23`? In that case this PR would probably have to be parked (once all the requested changes are made and approvals are given) until the library is built with `-std=c++26`.

Zingam replied:
@smallp-o-p You can check my attempt at `<debugging>`: https://github.com/llvm/llvm-project/pull/81447/files It is unfinished but it may give you some ideas and it builds. I think you also need to implement the AVAILABILITY macros.

frederick-vs-ja replied:
The library needs to be built with `-std=c++26` (or in an even later mode) as long as `text_encoding::environment` is separately compiled. So it doesn't make sense to guard this with `_LIBCPP_STD_VER >= 26`.

Alternatively, it's possible to provide an internal function achieving this functionality that's separately compiled and available in old modes. But it still makes no sense to guard the function definition in `src/`.

cor3ntin commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+    return text_encoding(__GNUC_EXECUTION_CHARSET_NAME);
+#      elif defined(__clang_literal_encoding__)
+    return text_encoding(__clang_literal_encoding__);
+#      elif defined(__clang__)
```
I suspect not, depends how many older versions of clang you support 
this macro was added in https://reviews.llvm.org/D100346?id=337008

Zingam commented on libcxx/test/std/utilities/text_encoding/text_encoding.ctor/id.pass.cpp:
```diff
@@ -0,0 +1,56 @@
+// 3. Constructing an object using id::unknown must set mib() to id::unknown and the name to an empty string.
+// 4. Constructing an object using id::other must set mib() to id::other and the name to an empty string.
+
+#include "test_text_encoding.h"
+#include <cassert>
+#include <string_view>
+#include <text_encoding>
+#include <type_traits>
```
```suggestion
#include <cassert>
#include <string_view>
#include <text_encoding>
#include <type_traits>

#include "test_text_encoding.h"
```

Nit: The common practice in tests is first Standard library headers, an empty line, and then any helper headers.

Zingam commented on libcxx/test/std/utilities/text_encoding/text_encoding.ctor/string_view.pass.cpp:
```diff
@@ -0,0 +1,73 @@
+
+// text_encoding::text_encoding(string_view) noexcept
+
+#include "test_macros.h"
+#include "test_text_encoding.h"
+#include <cassert>
+#include <cstring>
+#include <string_view>
+#include <text_encoding>
+#include <type_traits>
```
```suggestion
#include <cassert>
#include <cstring>
#include <string_view>
#include <text_encoding>
#include <type_traits>

#include "test_macros.h"
#include "test_text_encoding.h"
```

Same as above.

Zingam commented on libcxx/test/support/test_text_encoding.h:
```diff
@@ -0,0 +1,1173 @@
+#ifndef SUPPORT_TEST_TEXT_ENCODING_H
+#define SUPPORT_TEST_TEXT_ENCODING_H
+
+#include "test_macros.h"
+#include <cstdint>
+
```
```suggestion
#include <cstdint>

#include "test_macros.h"
```

Do you really need this header in the support directory? Perhaps it can be put in the test directory?

smallp-o-p commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+      {0, nullptr} // sentinel
+  };
+
+  const __encoding_data* __encoding_rep_ = __text_encoding_data + 1;
+  char __name_[max_name_length + 1]      = {0};
```
I don't think this is feasible due to [2.2](https://eel.is/c++draft/text.encoding.class#text.encoding.members-2.2) in the case of a match found for the name passed in `enc`, we'd have to copy the name into the buffer and somehow be able to retrieve the `id` _without_ the pointer.

frederick-vs-ja commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+  size_t operator()(const text_encoding& __enc) const noexcept { return std::hash<text_encoding::id>()(__enc.mib()); }
+};
+
+namespace ranges {
+
+template <>
+inline constexpr bool enable_borrowed_range<text_encoding::aliases_view> = true;
+
+} // namespace ranges
```
It seems better not to reopen the inner namespace.
```suggestion
template <>
inline constexpr bool ranges::enable_borrowed_range<text_encoding::aliases_view> = true;
```

frederick-vs-ja commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1482 @@
+  char __name_[max_name_length + 1]      = {0};
+};
+
+template <>
+struct hash<text_encoding> {
+  size_t operator()(const text_encoding& __enc) const noexcept { return std::hash<text_encoding::id>()(__enc.mib()); }
+};
```
Will it be valuable to put the `hash` specialization into another granularized internal header? @philnik777 

I think this is possibly the only cause suitable for granularization, as the `enable_borrowed_range` specialization should be available together with `aliases_view` to avoid potential ODR issues.

Zingam commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,70 @@
+#  pragma GCC system_header
+#endif
+
+#if _LIBCPP_STD_VER >= 26
+#  include <__text_encoding/text_encoding.h>
+#endif // _LIBCPP_STD_VER >= 26
```
I'm not arguing. IMO Without anything granularized it's just unnecessary overhead. I wondered why you didn't comment on it.

Zingam commented:
@smallp-o-p I think you might need to implement availability macros for macOS. I've done that just for free functions, so I'm not entirely sure. Look for the `availability` header.

smallp-o-p commented:
The issue with `text_encoding::environment()` is that the `text_encoding` class isn't visible in C++23, and I don't see how it would be possible to make it visible prior to C++26 other than:

1. Backport <text_encoding> to earlier versions of the library
2. Wait until libcxx is built with `-std=c++26` by default 

I'll have to dig in more with the availability macros, how `<filesystem>` was implemented seems promising.

frederick-vs-ja commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1460 @@
+
+  // [text.encoding.aliases], class text_encoding::aliases_view
+  struct aliases_view : ranges::view_interface<aliases_view> {
+    constexpr aliases_view() = default;
```
Do we need this default constructor? It's clarified in https://eel.is/c++draft/text.encoding.aliases#note-1 that `aliases_view` can be non-`default_initializable`.

smallp-o-p replied:
We don't, I should remove that.

frederick-vs-ja commented on libcxx/include/__text_encoding/text_encoding.h:
```diff
@@ -0,0 +1,1460 @@
+  // [text.encoding.aliases], class text_encoding::aliases_view
+  struct aliases_view : ranges::view_interface<aliases_view> {
+    constexpr aliases_view() = default;
+    constexpr aliases_view(const __encoding_data* __d) : __view_data_(__d) {}
```
This constructor seems too permissive (ditto `__iterator`'s), e.g. one can write `aliases_view(nullptr)`. I guess it's better to make it hard or even impossible to call constructor without mentioning any reserved identifier.

smallp-o-p replied:
`aliases_view(nullptr)` would result in an invalid iterator, however this should be a `private` constructor.

frederick-vs-ja commented:
> The issue with `text_encoding::environment()` is that the `text_encoding` class isn't visible in C++23, and I don't see how it would be possible to make it visible prior to C++26 other than:
> 
> 1. Backport `<text_encoding>` to C++23
> 2. Wait until libcxx is built with `-std=c++26` by default
> 
> I'll have to dig in more with the availability macros, how `<filesystem>` was implemented seems promising.

I think there's another approach. Note that `text_encoding` is a trivially copyable class (although not yet guarantee by the standard), we can define
- a `__text_encoding_rep` class which has the same layout as `text_encoding`, and
- a `__text_encoding_environment_rep` function returning it, and then
- use `std::bit_cast<text_encoding>(__text_encoding_environment_rep())` in `text_encoding::environment`.

The helper class and function can be available in old modes without exposing `text_encoding`. This approach can also make `text_encoding::environment` away from ABI boundary and potentially easier to evolve.

cor3ntin commented:
I have no idea what libc++ policies are in terms of backporting, but given that this feature is most useful for legacy systems, it might be reasonable to make it available in C++20 (we need consteval) - assuming there are appropriate warnings.

Zingam commented:
> * Android 5.0 and 13.0 don't have `nl_langinfo_l`, but do have `<langinfo.h>` (~on further research, it looks like that shouldn't even be possible, `<langinfo.h>` is not part of the Android NDK...~ I've found some bits of source code of the NDK that have it, some that don't. Will have to investigate what NDK the CI uses. From the bits I've found, it's guarded by `!__LP64__`...)
> * There is a macro collision from `<langinfo.h>` in two of the tests

I belive the latest released NDK.

Zingam commented on libcxx/test/std/utilities/text_encoding/text_encoding.members/text_encoding.aliases_view/begin.pass.cpp:
```diff
@@ -0,0 +1,65 @@
+
+// REQUIRES: std-at-least-c++26
+
+// struct text_encoding::aliases_view
```
```suggestion
// REQUIRES: std-at-least-c++26

// <text_encoding>

// struct text_encoding::aliases_view
```

We put the `REQUIRES`, `UNSUPPORTED`... etc. on top and then the synopsis: the header name, etc.

Zingam commented on libcxx/test/std/utilities/text_encoding/text_encoding.members/text_encoding.aliases_view/begin.pass.cpp:
```diff
@@ -0,0 +1,65 @@
+  {
+    auto te    = std::text_encoding(id::UTF8);
+    auto view1 = te.aliases();
+    auto view2 = te.aliases();
```
AFAIK It is recommended to specify the full type and avoid using `auto` if possible.

H-G-Hristov commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,142 @@
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit text_encoding(string_view __enc) noexcept : __impl_(__enc) {}
+  _LIBCPP_HIDE_FROM_ABI constexpr text_encoding(id __i) noexcept : __impl_(__i) {}
+
+  [[__nodiscard__]] _LIBCPP_HIDE_FROM_ABI constexpr id mib() const noexcept { return __impl_.__mib(); }
```
```suggestion
  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr id mib() const noexcept { return __impl_.__mib(); }
```
This should be consistently `[[nodiscard]]`.

H-G-Hristov commented on libcxx/include/__text_encoding/te_impl.h:
```diff
@@ -0,0 +1,1439 @@
+_LIBCPP_PUSH_MACROS
+#include <__undef_macros>
+
+#if _LIBCPP_STD_VER >= 23
```
Why is this "23"?

H-G-Hristov commented on libcxx/test/std/localization/locale.categories/category.monetary/locale.money.get/locale.money.get.members/get_long_double_fr_FR.pass.cpp:
```diff
@@ -33,9 +33,9 @@
 #ifdef _AIX
 // the AIX libc expects U202F as LC_MONETARY thousands_sep
-#  define THOUSANDS_SEP L"\u202F"
+#  define THOUSANDS_SEP_ L"\u202F"
```
Why did you make this change? It seems unrelated and unnecessary? Unrated changes should be done in separate patches.

smallp-o-p commented on libcxx/test/std/localization/locales/locale/locale.members/encoding.pass.cpp:
```diff
@@ -0,0 +1,53 @@
+using id = std::text_encoding::id;
+
+int main(int, char**) {
+#if false
```
Test for `locale::encoding()`, which I don't envision being able to implement in this PR. Wasn't sure whether to remove it or leave it in for the future. I do remember seeing similar tests for `optional<T&>` that were commented out like this (albeit those were parts of a greater test), but that was likely 8 years ago...so policy regarding stuff like this is different.

smallp-o-p committed:
Retarget to LLVM 23

smallp-o-p committed:
Improve struct size, 72 bytes -> 66 bytes on 64bit system

smallp-o-p committed:
Update locale for AIX, picolibc

smallp-o-p committed:
Reduce __te_data struct size from 16B -> 6B, reduces data table size from
14144 Bytes -> 8620 Bytes

smallp-o-p committed:
Clean up ctor tests

smallp-o-p committed:
Remove wrap_iter change

smallp-o-p committed:
Add more GCC steps

smallp-o-p commented on libcxx/include/__text_encoding/te_impl.h:
```diff
@@ -0,0 +1,1492 @@
+  // implementation simple.
+
+  static constexpr __te_data __entries[] = {
+      {0, 1, 0},      {1, 2, 0},      {2, 3, 11},     {13, 4, 9},     {22, 5, 7},     {29, 6, 7},     {36, 7, 7},
```
[Size comparison between `libstdc++` implementation and this one, for reference](https://godbolt.org/z/n47qvqvv8)

cor3ntin replied:
This could probably benefit from a comment

smallp-o-p replied:
[Even more savings!](https://godbolt.org/z/b81z96h3o)

ldionne commented on libcxx/include/__text_encoding/te_impl.h:
```diff
@@ -0,0 +1,1492 @@
+  static constexpr size_t __max_name_length_ = 63;
+  struct __te_data {
+    unsigned short __first_alias_index_;
+    unsigned short __mib_rep_;
```
What do you think about using the enumerator directly here?

ldionne commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,142 @@
+
+#if __cplusplus < 201103L && defined(_LIBCPP_USE_FROZEN_CXX03_HEADERS)
+#  include <__cxx03/__config>
+#else
```
I suspect we can just remove this since this is a C++26 only header.

ldionne commented on libcxx/test/libcxx/transitive_includes/cxx03.csv:
```diff
@@ -2282,6 +2282,39 @@ system_error typeinfo
 system_error utility
 system_error variant
 system_error version
+text_encoding algorithm
```
We probably want to make the header empty prior to C++26, i.e. not include anything from it. That would imply doing:

```
#include <__config>
#if C++26

#  include <other headers>

contents

#endif // C++26
```

I think we do that in a few places. I seem to recall we once had a push to do that for all of our headers, but I don't recall what happened.

ldionne reviewed (commented):
Thanks a lot for the patch! This is a first pass.

smallp-o-p commented on libcxx/include/__text_encoding/te_impl.h:
```diff
@@ -0,0 +1,1492 @@
+      {857, 2108, 2}, {859, 2109, 2}, {861, 2250, 2}, {863, 2251, 2}, {865, 2252, 2}, {867, 2253, 2}, {869, 2254, 2},
+      {871, 2255, 2}, {873, 2256, 2}, {875, 2257, 2}, {877, 2258, 2}, {879, 2259, 3}, {882, 2260, 2}};
+
+  static constexpr char const* __aliases_table[] = {
```
[A surprisingly eye-watering amount of savings!](https://godbolt.org/z/eM17WajGj)

Edit: But it may make `aliases_view` a bit more complicated.

smallp-o-p committed:
Clean up header

smallp-o-p committed:
Add enumeration test

smallp-o-p committed:
Fix iterator

smallp-o-p committed:
Add noexcept tests

smallp-o-p committed:
Strengthen iterator tests

smallp-o-p committed:
Update transitive includes

smallp-o-p committed:
Formatting + missed test change

smallp-o-p committed:
Fix test failures

smallp-o-p committed:
Move version header

ldionne commented on libcxx/include/__locale_dir/locale_base_api.h:
```diff
@@ -157,6 +158,7 @@ namespace __locale {
 #    define _LIBCPP_MESSAGES_MASK LC_MESSAGES_MASK
 #    define _LIBCPP_ALL_MASK LC_ALL_MASK
 #    define _LIBCPP_LC_ALL LC_ALL
+#    define _LIBCPP_NL_CODESET CODESET
```
We need to update the documentation above for the API with this new macro and with the new function (`nl_langinfo`).

Sharjeel-Khan commented on libcxx/src/text_encoding.cpp:
```diff
@@ -0,0 +1,256 @@
+}
+
+#else
+_LIBCPP_HIDDEN __te_impl __te_impl::__get_locale_encoding(const char* __name [[maybe_unused]]) { return __te_impl(); }
```
I asked Elliott and Android's Bionic does implement `nl_langinfo_l` and `nl_langinfo`. We do not know how the assumption of 32-bit came but it should be available on both. You can find both `nl_langinfo_l` and `nl_langinfo` implemented in Bionic as seen here: https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/langinfo.cpp. It was added in API 26 so there should be an API check: `if (defined(__ANDROID__) && __ANDROID_API__ >= 26)`.

smallp-o-p committed:
Make environment helpers friend functions

smallp-o-p committed:
[AI-Assisted] Implement __get_locale_encoding on Windows

smallp-o-p committed:
Deuglify nodiscards, remove on deleted functions

smallp-o-p committed:
Bump step count

smallp-o-p committed:
Update C locale API and documentation, enable for Android API >= 26

smallp-o-p committed:
Format

smallp-o-p committed:
Gate environment() on android

smallp-o-p committed:
Android default encoding is UTF-8

smallp-o-p committed:
Fix Windows build

smallp-o-p committed:
ditto

smallp-o-p committed:
ditto

smallp-o-p committed:
ditto

smallp-o-p committed:
Fix Windows implementation

smallp-o-p committed:
Don't rely on system-specific encodings in environment test

smallp-o-p committed:
Store aliases as one string and store a table of indexes into that string
Reduces runtime size 18.4Ki -> 13.1Ki
Reduces file size 41.4Ki -> 15.5Ki
(mostly attributed to eliminated relocation data)

smallp-o-p committed:
Formatting

smallp-o-p committed:
Revert "Don't rely on system-specific encodings in environment test"

This reverts commit f69991322d5f264c8caffdda59a9d4ee33ef3179.

smallp-o-p committed:
Make environment failure message a little easier

smallp-o-p committed:
Assume android to be UTF-8 always

smallp-o-p committed:
Fix test

smallp-o-p commented on libcxx/test/std/utilities/text_encoding/text_encoding.members/environment.pass.cpp:
```diff
@@ -0,0 +1,61 @@
+  // 1. Depending on the platform's default, verify that environment() returns the corresponding text encoding.
+  {
+#  if defined(__ANDROID__)
+    assert(te.mib() == std::text_encoding::UTF8);
```
> for what purpose, exactly? what can you do with that information?

`locale::encoding()`: "Returns a [std::text_encoding](https://en.cppreference.com/w/cpp/locale/text_encoding.html) object that represents the implementation-defined encoding scheme associated with the locale `*this`."

From what I'm understanding, the set of valid `std::locale` objects is essentially limited to:

- ""
- "C"
- "C.UTF-8"
- "en_US.UTF-8"
- "POSIX"

and they're all associated with UTF-8. Otherwise, we can just return unknown, likely because someone is trying to ask about a locale the system doesn't support. Is this correct?

smallp-o-p committed:
Implement __get_locale_encoding(..) for Android

ldionne commented on libcxx/test/std/utilities/text_encoding/text_encoding.members/text_encoding.aliases_view/begin.pass.cpp:
Please use a full synopsis. What does this return? You can use `implementation-defined`.

smallp-o-p committed:
Merge remote-tracking branch 'upstream/main' into 105373-implement-P1885R12-P2862R1

smallp-o-p committed:
Library now built with C++26

smallp-o-p committed:
Address test comments

smallp-o-p committed:
Implement locale::encoding()

smallp-o-p committed:
Typo

smallp-o-p committed:
Expose only internal __get_locale_encoding(...)

smallp-o-p committed:
ABI list, transitive includes,

smallp-o-p committed:
More test adjustments

smallp-o-p committed:
abi-list, fix includes, add debug print for encoding.pass.cpp

smallp-o-p committed:
Formatting, papers

smallp-o-p committed:
Expose only __get_locale_encoding

smallp-o-p committed:
Remove default locale encoding test

smallp-o-p committed:
Fix GCC, update ABI list

smallp-o-p committed:
Fix tests

smallp-o-p committed:
Cleanup

smallp-o-p committed:
ci

smallp-o-p committed:
apple abi

smallp-o-p committed:
Return UTF-8 always on Android, try parse the locale name on Windows

smallp-o-p committed:
Fix Windows, locale::encoding() availability

smallp-o-p committed:
fix test failures

smallp-o-p committed:
accidentally committed changes

ldionne commented on libcxx/include/__configuration/availability.h:
```diff
@@ -70,8 +72,10 @@
+#  define _LIBCPP_INTRODUCED_IN_LLVM_23_ATTRIBUTE __attribute__((unavailable))
+
 // LLVM 22
-// TODO: Fill this in
```
This should not be removed. We still need to fill this in for Apple.

ldionne commented on libcxx/include/__locale:
```diff
@@ -107,6 +109,12 @@ public:
+#  if _LIBCPP_STD_VER >= 26
+  _LIBCPP_AVAILABILITY_TE_ENVIRONMENT _LIBCPP_HIDE_FROM_ABI std::text_encoding encoding() const {
+    std::string __name = name();
+    return __get_locale_encoding(__name.c_str());
```
```suggestion
    return std::__get_locale_encoding(__name.c_str());
```

smallp-o-p committed:
Address non-code related comments

smallp-o-p committed:
Generate offsets at compile-time

smallp-o-p committed:
Documentation

smallp-o-p committed:
Merge remote-tracking branch 'upstream/main' into 105373-implement-P1885R12-P2862R1

smallp-o-p committed:
ABI annotations

smallp-o-p committed:
Test every string in database

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/default.pass.cpp:
```diff
@@ -0,0 +1,42 @@
+//===----------------------------------------------------------------------===//
```
I guess it's better to place the test files in the `‎libcxx/test/std/text/text_encoding` subdirectory (or `‎libcxx/test/std/text/text.encoding` which is more consistent with the standard section label).

smallp-o-p committed:
Move tests under test/{libcxx,std}/text/

frederick-vs-ja commented on libcxx/include/__locale:
```diff
@@ -108,6 +110,12 @@ public:
                __x.data(), __x.data() + __x.size(), __y.data(), __y.data() + __y.size()) < 0;
   }
+#  if _LIBCPP_STD_VER >= 26
+  _LIBCPP_AVAILABILITY_TE_ENVIRONMENT _LIBCPP_HIDE_FROM_ABI std::text_encoding encoding() const {
```
```suggestion
  [[nodiscard]] _LIBCPP_AVAILABILITY_TE_ENVIRONMENT _LIBCPP_HIDE_FROM_ABI std::text_encoding encoding() const {
```

Now we've added `[[nodiscard]]` to `<locale>`. I think it's time to add `[[nodiscard]]` and enable test coverage in this line:
https://github.com/llvm/llvm-project/blob/e2efa2e958270e0f21d9db2aa9a393006c9d6d0b/libcxx/test/libcxx/localization/nodiscard.verify.cpp#L33

frederick-vs-ja commented on libcxx/include/__locale:
```diff
@@ -108,6 +110,12 @@ public:
   }
+#  if _LIBCPP_STD_VER >= 26
+  _LIBCPP_AVAILABILITY_TE_ENVIRONMENT _LIBCPP_HIDE_FROM_ABI std::text_encoding encoding() const {
+    std::string __name = name();
+    return std::__get_locale_encoding(__name.c_str());
```
It's a bit silly to copy the string as the string is already stored internally.

https://github.com/llvm/llvm-project/blob/e2efa2e958270e0f21d9db2aa9a393006c9d6d0b/libcxx/src/locale.cpp#L116

I _think_ we can use some technique to obtain the `c_str()` within the separately compiled library (e.g. making `__get_locale_encoding` take `const locale::__imp*` instead), although I'm not sure whether such approach is preferred.

smallp-o-p replied:
I'm not sure it's a good idea to tie `__get_locale_encoding()` to the `std::locale` internal implementation. Another possible way we could do this is introducing a private `__encoding()` method that's implemented in a `.cpp` file, where we know the implementation of `locale::__imp`, but that would require exporting another symbol.

I think it can be done in a future change, as I'd like to keep the amount of symbols introduced small for this one.

frederick-vs-ja replied:
Hmm, then how about taking ~`const locale&`~ `const locale*` instead? I think it's still possible to export only one symbol while avoiding copying the string.

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,832 @@
+      using iterator_category = random_access_iterator_tag;
+      using value_type        = const char*;
+      using reference         = const char*;
+      using difference_type   = decltype(static_cast<unsigned short>(1) - static_cast<unsigned short>(1));
```
Current `difference_type` is basically `int` on all platforms we support. I personally don't want to see such confusing spelling.
```suggestion
      using difference_type   = int;
```

Also, the current technique will give incorrect `unsigned int` on 16-bit LP32 platforms, although libc++ doesn't support them currently.

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,832 @@
+
+      constexpr __iterator() noexcept = default;
+
+      constexpr reference operator*() const noexcept { return __get_alias_from_offset_table(__current_idx_); }
```
This should also be `[[nodiscard]]`.
```suggestion
      [[nodiscard]] constexpr reference operator*() const noexcept { return __get_alias_from_offset_table(__current_idx_); }
```

Same for `operator[]`, `operator+`, and `operator-`, but not comparison operators.

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,832 @@
+
+    constexpr __iterator begin() const noexcept { return __iterator(__first_idx_); }
+
+    constexpr __iterator end() const noexcept { return __iterator(__last_idx_); }
```
```suggestion
    [[nodiscard]] constexpr __iterator begin() const noexcept { return __iterator(__first_idx_); }

    [[nodiscard]] constexpr __iterator end() const noexcept { return __iterator(__last_idx_); }
```

frederick-vs-ja commented on libcxx/test/libcxx/text/text_encoding/text_encoding.members/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,31 @@
+//===----------------------------------------------------------------------===//
```
I think it's simpler to have only one `libcxx/test/libcxx/text/text_encoding/nodiscard.verify.cpp` instead and verify `alias_view::begin`/`end`, `hash::operator()`, and iterator operations together.

For `environment` and `environment_is`, we can guard the statements with `#ifndef TEST_HAS_NO_LOCALIZATION`.

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,832 @@
+        return *(__it + __n);
+      }
+
+      constexpr decltype(auto) operator->() = delete;
```
Is this helpful? I don't think we need to declare `operator->` at all.

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,832 @@
+
+template <>
+struct hash<text_encoding> {
+  size_t operator()(const text_encoding& __enc) const noexcept {
```
```suggestion
  [[nodiscard]] size_t operator()(const text_encoding& __enc) const noexcept {
```

ldionne commented on libcxx/include/__configuration/availability.h:
```diff
@@ -289,6 +296,10 @@
 
+// Controls whether the implementation for text_encoding::environment() is available
+#define _LIBCPP_AVAILABILITY_HAS_TE_ENVIRONMENT _LIBCPP_INTRODUCED_IN_LLVM_23
+#define _LIBCPP_AVAILABILITY_TE_ENVIRONMENT _LIBCPP_INTRODUCED_IN_LLVM_23_ATTRIBUTE
```
What would you think of `_LIBCPP_AVAILABILITY_HAS_TEXT_ENCODING_ENVIRONMENT`?

ldionne commented on libcxx/lib/abi/i686-linux-android23.libcxxabi.v1.stable.exceptions.nonew.abilist:
```diff
@@ -2335,4 +2336,4 @@
 {'is_defined': True, 'name': '__cxa_vec_new2', 'type': 'FUNC'}
-{'is_defined': True, 'name': '__cxa_vec_new3', 'type': 'FUNC'}
\ No newline at end of file
+{'is_defined': True, 'name': '__cxa_vec_new3', 'type': 'FUNC'}
```
Let's drop these newline changes from this patch. We should add a newline separately and also make sure that the scripts that produce these files in the CI do include a newline. That way everything's in sync. But not as part of this patch.

ldionne commented on libcxx/test/libcxx/text/text_encoding/text_encoding.ctor/assert.id.pass.cpp:
```diff
@@ -0,0 +1,32 @@
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: std-at-least-c++26
+// REQUIRES: libcpp-hardening-mode={{extensive|debug}}
```
I think you also need `has-unix-headers` or something like that, please take a look at the other assertion tests. Apply throughout your added assertion tests.

ldionne commented on libcxx/test/libcxx/text/text_encoding/text_encoding.members/environment.delete.verify.cpp:
```diff
@@ -0,0 +1,29 @@
+
+#include <text_encoding>
+
+int main(int, char**) {
```
For `verify` tests, we normally don't use `main()` functions since they're not necessary, to avoid giving the false impression that the test runs. Instead, we use e.g. `void f()` or `void test()`.

ldionne commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,832 @@
+
+_LIBCPP_END_EXPLICIT_ABI_ANNOTATIONS
+
+struct text_encoding {
```
You mentioned that we're assuming `CHAR_BITS == 8` throughout this file but the spec technically supports other values for it. Let's add a static assertion to document this assumption.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/string_view.pass.cpp:
```diff
@@ -0,0 +1,110 @@
+  for (auto& enc : all_encoding_data) {
+    test_ctor(enc.name, std::text_encoding::id(enc.mib));
+  }
+}
```
I'd inline this function into `main`.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/string_view.pass.cpp:
```diff
@@ -0,0 +1,110 @@
+
+constexpr bool test() {
+  // happy paths
+  {
```
These local scopes are not really buying us anything, I'd suggest removing them.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/string_view.pass.cpp:
```diff
@@ -0,0 +1,110 @@
+  }
+}
+
+constexpr void test_others() {
```
I'm not strongly attached, but I'd suggest also inlining this function and `test_primary_encoding_spellings` since they're only called once.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.eq/equal.id.pass.cpp:
```diff
@@ -0,0 +1,75 @@
+    assert(std::text_encoding(id::UTF8) == id::UTF8);
+  }
+
+  // 2.1
```
IMO this sub-numbering is a bit confusing.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.eq/equal.pass.cpp:
```diff
@@ -0,0 +1,69 @@
+    std::text_encoding other_te2 = std::text_encoding("bar");
+    assert(!(other_te1 == other_te2));
+  }
+
```
Here, I would add a single test that does something like `(a == b) == !(a != b)`, just to sanity check that we don't do anything that would result in e.g. the `operator!=` being deleted/disabled/not-provided.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.aliases_view/begin.pass.cpp:
```diff
@@ -0,0 +1,65 @@
+
+// constexpr implementation-defined text_encoding::aliases_view::begin() const;
+
+#include "test_macros.h"
```
This should go after the libc++ includes.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.aliases_view/index.pass.cpp:
```diff
@@ -0,0 +1,37 @@
+//===----------------------------------------------------------------------===//
```
Test file name is wrong.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.aliases_view/empty.pass.cpp:
```diff
@@ -0,0 +1,56 @@
+
+// <text_encoding>
+
+// text_encoding::aliases_view
```
This synopsis is incomplete.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.aliases_view/end.pass.cpp:
```diff
@@ -0,0 +1,37 @@
+  // 1. begin() of an aliases_view of "other" is equal to end()
+  ASSERT_NOEXCEPT(other_aliases.end());
+  assert(other_aliases.begin() == other_aliases.end());
+
```
I think we should also have a test to ensure that if we do have aliases, then `begin() != end()`.

Potentially also that incrementing `begin()` successively eventually reaches `end()` for such a view -- but that would belong in another test file (maybe `begin()`?).

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.aliases_view/operator-bool.pass.cpp:
```diff
@@ -0,0 +1,60 @@
+
+// <text_encoding>
+
+// text_encoding::aliases_view
```
Incomplete synopsis.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/aliases_view.compile.pass.cpp:
```diff
@@ -0,0 +1,24 @@
+static_assert(std::ranges::borrowed_range<std::text_encoding::aliases_view>);
+static_assert(std::same_as<std::ranges::range_value_t<std::text_encoding::aliases_view>, const char*>);
+static_assert(std::same_as<std::ranges::range_reference_t<std::text_encoding::aliases_view>, const char*>);
+static_assert(noexcept(std::text_encoding().aliases()));
```
We should have an explicit test file for `text_encoding::aliases()`. That's where you'd ensure the existence of the function, its return type, constexpr friendliness, etc. And then this `noexcept` test should go away.

ldionne commented on libcxx/test/std/text/text_encoding/text_encoding.members/id.compile.pass.cpp:
```diff
@@ -0,0 +1,276 @@
+static_assert(int_least32_t(std::text_encoding::id::EBCDICES) == 2074);
+static_assert(int_least32_t(std::text_encoding::id::EBCDICESA) == 2075);
+static_assert(int_least32_t(std::text_encoding::id::EBCDICESS) == 2076);
+static_assert(int_least32_t(std::text_encoding::id::EBCDICUK) == 2077);
```
This is going to be very pedantic, but we need this to be a runtime test (with the usual constexpr / non constexpr pattern).

Otherwise, you can `static_assert(std::is_enum_v<decltype(std::text_encoding::id)>);` and I'd be happy too, since you'd be pinning the fact that these are enum constants.

ldionne commented on libcxx/test/std/text/text_encoding/trivially_copyable.compile.pass.cpp:
```diff
@@ -0,0 +1,16 @@
+
+#include <text_encoding>
+#include <type_traits>
+
```
I'd suggest we add a comment saying that the spec requires it (it's not suuuuper frequent so it might otherwise be thought to be a libc++ specific property that we're incorrectly testing here).

ldionne reviewed (approved):
Lots of small comments. As far as I'm concerned, this can be shipped once my comments are applied and @frederick-vs-ja 's too. I'll let @frederick-vs-ja give the final approval once his comments are satisfied.

smallp-o-p committed:
Merge remote-tracking branch 'upstream/main' into 105373-implement-P1885R12-P2862R1

smallp-o-p committed:
Address most comments 1

smallp-o-p committed:
Address review comments 2

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,835 @@
+
+private:
+  static constexpr int __NATS_DANO_     = 33;
+  static constexpr int __NATS_DANO_ADD_ = 34;
```
It seems that these internal names will easily conflict with internal _UGLY macro names. I think it would be better to use __ugly_lower_snake_case and move them into `__find_data_idx_by_id`.

frederick-vs-ja commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,835 @@
+
+template <>
+struct hash<text_encoding> {
+  [[nodiscard]] size_t operator()(const text_encoding& __enc) const noexcept {
```
Oh, I'm sorry for overlooking another improvement here. Perhaps it would be a bit better (for performance) to use `static` here.

```suggestion
  [[nodiscard]] static size_t operator()(const text_encoding& __enc) noexcept {
```

frederick-vs-ja commented on libcxx/test/libcxx/text/text_encoding/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,67 @@
+
+// <text_encoding>
+
+#include <functional>
```
I don't think we need this. Including `<text_encoding>` should be sufficient for obtaining `std::hash`.
If we find that such removal breaks Clang module builds, we should fix `module.modulemap.in‎` by adding `export std.functional.hash` to the `text_encoding` module (see also #169634 and #170674).

```suggestion
```

frederick-vs-ja commented on libcxx/test/libcxx/text/text_encoding/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,67 @@
+
+#include "test_macros.h"
+
+int main(int, char**) {
```
We should change this to
```C++
void test() {
  // ...
  // no `return 0;` here
}
```

frederick-vs-ja commented on libcxx/test/libcxx/text/text_encoding/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,67 @@
+#include "test_macros.h"
+
+int main(int, char**) {
+  std::text_encoding te = std::text_encoding();
```
I think we should use either default-initialization (`std::text_encoding te;`) or empty list-initialization (`std::text_encoding te{};` or `std::text_encoding te = {};`) for clarity. If you want to use value-initialization, we may also use `auto te = std::text_encoding();` to avoid typing the same type-id twice.

Same for all similar `std::text_encoding` variable definitions in test files.

frederick-vs-ja commented on libcxx/test/libcxx/text/text_encoding/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,67 @@
+
+  // Clang does not emit a nodiscard warning for consteval functions with [[nodiscard]]: See issue #141536
+  // expected-warning@+1 {{expression result unused}}
+  std::text_encoding::literal();
```
I think we can move this line to a separated `consteval` function and verify the warning message in the usual way. I used such technique in #200760.

smallp-o-p replied:
Neat, thanks!

frederick-vs-ja commented on libcxx/test/std/localization/locales/locale/locale.members/encoding.pass.cpp:
```diff
@@ -0,0 +1,52 @@
+  {
+    // 1. Locale built with en_US.UTF-8 returns text_encoding representing "UTF-8"
+    const std::locale utf8_locale(LOCALE_en_US_UTF_8);
+    std::text_encoding te      = utf8_locale.encoding();
```
I think in this line it would make sense to use the `std::same_as<...> decltype(auto)` technique. (`<concepts>` should be included if use this.)

frederick-vs-ja commented on libcxx/test/std/localization/locales/locale/locale.members/encoding.pass.cpp:
```diff
@@ -0,0 +1,52 @@
+    // 1. Locale built with en_US.UTF-8 returns text_encoding representing "UTF-8"
+    const std::locale utf8_locale(LOCALE_en_US_UTF_8);
+    std::text_encoding te      = utf8_locale.encoding();
+    std::text_encoding utf8_te = std::text_encoding(std::text_encoding::id::UTF8);
```
In any case, I don't think it's a good idea to type `std::text_encoding` so many times in this variable definition.

I in this line it would be clearer to use either
- `auto te = std::text_encoding(std::text_encoding::id::UTF8);` or
- `auto te = std::text_encoding{std::text_encoding::id::UTF8};`.

For all similar variable definitions in test files, we may also use
- `std::text_encoding te{std::text_encoding::id::UTF8};`,
- `std::text_encoding te = {std::text_encoding::id::UTF8};`, or
- `std::text_encoding te = std::text_encoding::id::UTF8;`.

Note that `id::` can be omitted as `id` is `using enum`'d. Although it might be clearer to keep it.

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/default.pass.cpp:
```diff
@@ -0,0 +1,42 @@
+#include <type_traits>
+
+constexpr bool test() {
+  std::same_as<std::text_encoding> decltype(auto) te = std::text_encoding();
```
In contrast to testing `locale::encoding`, I think it's not so meaningful to verity the type sameness, because we're explicitly typing `std::text_encoding` in the initializer.

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/default.pass.cpp:
```diff
@@ -0,0 +1,42 @@
+  {
+    test();
+    static_assert(test());
+  }
```
Let's use `_v` and avoid pointless `{}` (throughout all new test files). I think additional braces are only meaningful when there're local variables.
```suggestion
  // 1. Default constructor must be nothrow
  static_assert(std::is_nothrow_default_constructible_v<std::text_encoding>, "Must be nothrow default constructible");

  // 2. Default constructing a text_encoding object makes it so that mib() == id::unknown, and its name is empty
  test();
  static_assert(test());
```

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/id.pass.cpp:
```diff
@@ -0,0 +1,73 @@
+  {
+    test();
+    static_assert(test());
+  }
```
```suggestion
  // text_encoding(id) must be nothrow
  static_assert(std::is_nothrow_constructible_v<std::text_encoding, std::text_encoding::id>,
                "Must be nothrow constructible with id");

  test();
  static_assert(test());
```

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.ctor/string_view.pass.cpp:
```diff
@@ -0,0 +1,80 @@
+}
+
+constexpr bool test() {
+  { // The first encoding name for each mib in the data table.
```
Ditto here and below. `{}` is pointeless.

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.eq/equal.id.pass.cpp:
```diff
@@ -0,0 +1,61 @@
+constexpr bool test() {
+  // operator==(const text_encoding&, id) must be noexcept and returns bool
+  {
+    std::text_encoding te = std::text_encoding();
```
No change requested for the `{}` as there's a local variable. But as mentioned before, the style of initialization should be simplified.
```suggestion
  {
    std::text_encoding te{};
```

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.members/id.compile.pass.cpp:
```diff
@@ -0,0 +1,278 @@
+#include <text_encoding>
+#include <type_traits>
+
+static_assert(std::is_enum_v<std::text_encoding::id>);
```
```suggestion
static_assert(std::is_scoped_enum_v<std::text_encoding::id>);
static_assert(std::is_same_v<std::underlying_type_t<std::text_encoding::id>, std::int_least32_t>);
```

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.members/id.compile.pass.cpp:
```diff
@@ -0,0 +1,278 @@
+#include <type_traits>
+
+static_assert(std::is_enum_v<std::text_encoding::id>);
+static_assert(int_least32_t(std::text_encoding::id::other) == 1);
```
It would be a bit too pedantic, but the standard doesn't guarantee that `<cstdlint>` also provides `int_least32_t` in the global namespace. Pedantically, we should either use `using std::int_least32_t;` above or `std::`-qualify every occurrence of `int_least32_t` here and below.

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.consteval.verify.cpp:
```diff
@@ -0,0 +1,18 @@
+
+constexpr decltype(auto) foo() {
+  return &std::text_encoding::
+      literal; // expected-error@*:* {{cannot take address of consteval function 'literal' outside of an immediate invocation}}
```
Just a style change. Given there's `@*:*` (which expects the diagnose message to appear for an arbitrary line), we don't need to add the comment in the same line.
```suggestion
  // expected-error@*:* {{cannot take address of consteval function 'literal' outside of an immediate invocation}}
  return &std::text_encoding::literal; 
```

smallp-o-p committed:
Address remaining comments

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.members/literal.pass.cpp:
```diff
@@ -0,0 +1,36 @@
+#include <text_encoding>
+
+constexpr bool test() {
+  std::text_encoding te = std::text_encoding::literal();
```
The `std::same_as<...> decltype(auto)` technique would be meaningful here.

frederick-vs-ja commented on libcxx/test/libcxx/localization/encoding.nodiscard.verify.cpp:
```diff
@@ -0,0 +1,24 @@
+// UNSUPPORTED: availability-te-environment-missing
+
+// <locale>
+// In a separate file because the implementation for locale::encoding() is guarded by availability
```
Couldn't we move these to the main `nodicard.verify.cpp` file, and additionally guard the part with `!defined(_LIBCPP_VERSION) || defined(_LIBCPP_AVAILABILITY_TEXT_ENCODING_ENVIRONMENT)`?

Same for test coverage for `text_encoding::environment`/`text_encoding::environment_is`.

smallp-o-p committed:
Address comments 3, reduce the amount fo std::text_encoding type uses

smallp-o-p committed:
Formatting

smallp-o-p committed:
Fix nodiscard test

frederick-vs-ja commented on libcxx/test/std/text/text_encoding/text_encoding.members/literal.pass.cpp:
```diff
@@ -0,0 +1,36 @@
+// text_encoding text_encoding::literal() noexcept;
+
+#include <cassert>
+#include <text_encoding>
```
```suggestion
#include <concepts>
#include <text_encoding>
```

smallp-o-p committed:
Update libcxx/test/std/text/text_encoding/text_encoding.members/literal.pass.cpp

Co-authored-by: A. Jiang <de34@live.cn>

frederick-vs-ja commented on libcxx/test/libcxx/text/text_encoding/nodiscard.verify.cpp:
```diff
@@ -0,0 +1,64 @@
+  // expected-warning@+1 {{ignoring return value of function declared with 'nodiscard' attribute}}
+  std::hash<std::text_encoding>()(std::text_encoding::ASCII);
+
+#if _LIBCPP_AVAILABILITY_HAS_TEXT_ENCODING_ENVIRONMENT
```
`!defined(TEST_HAS_NO_LOCALIZATION)` should be sufficient for fixing the remaining failures...
```suggestion
#if !defined(TEST_HAS_NO_LOCALIZATION) && _LIBCPP_AVAILABILITY_HAS_TEXT_ENCODING_ENVIRONMENT
```

philnik777 commented on libcxx/include/__locale_dir/locale_base_api.h:
```diff
@@ -102,6 +103,8 @@
 //  int     __snprintf(char*, size_t, __locale_t, const char*, ...); // required by the headers
 //  int     __asprintf(char**, __locale_t, const char*, ...);        // required by the headers
+//
+// const char* __nl_langinfo(decltype(_LIBCPP_NL_CODESET) __item, __locale_t __loc);
```
I'm _really_ not convinced this is the right abstraction, given that at least Windows can't actually support this API. I'd _much_ rather have a more abstract function which then calls the appropriate system API than having something that mostly works.

frederick-vs-ja replied:
Yeah. IIUC we can just use `const char* __codeset = ::nl_langinfo(CODESET, __l);` in `libcxx/src/text_encoding.cpp‎` for simplicity.

philnik777 replied:
I think I'd rather go the other way and have something like `__get_locale_encoding` as part of the locale base API. There is a good chance that there is another system out there with different requirements, and I don't see much of a reason to not have this as part of the locale API.

smallp-o-p committed:
Fix test

smallp-o-p committed:
Merge remote-tracking branch 'refs/remotes/origin/105373-implement-P1885R12-P2862R1' into 105373-implement-P1885R12-P2862R1

smallp-o-p committed:
Create a function in the locale base API to query the encoding name of a locale

Without requiring nl_langinfo_l

smallp-o-p committed:
Move __make_text_encoding in else block

smallp-o-p committed:
Return text after dot in windows implementation

smallp-o-p committed:
Typo

smallp-o-p committed:
Merge remote-tracking branch 'upstream/main' into 105373-implement-P1885R12-P2862R1

philnik777 commented on libcxx/include/__locale_dir/support/bsd_like.h:
```diff
@@ -46,6 +47,7 @@ namespace __locale {
 #define _LIBCPP_MESSAGES_MASK LC_MESSAGES_MASK
 #define _LIBCPP_ALL_MASK LC_ALL_MASK
 #define _LIBCPP_LC_ALL LC_ALL
+#define _LIBCPP_NL_CODESET CODESET
```
I don't think this is required?

philnik777 commented on libcxx/test/std/text/text_encoding/text_encoding.members/text_encoding.consteval.verify.cpp:
```diff
@@ -0,0 +1,18 @@
+
+constexpr decltype(auto) foo() {
+  // expected-error@*:* {{cannot take address of consteval function 'literal' outside of an immediate invocation}}
+  return &std::text_encoding::literal;
```
This is definitely non-conforming code - you're taking the address of a standard library function. Is there any other way to check that something is `consteval`? If not, this is an implementation detail and the guideline is wrong.

smallp-o-p replied:
As far as I can tell this trick is the only way we can test that something is *always* a constant expression, and not just 'can be run as a constant expression.' I'll drop the test and guideline because I'm not sure of another way.

cor3ntin replied:
The language ill-formedness takes precedence over the library UB and associated warning so from a clang perspective i think this is fine to have in a test - Does it make sense to test constevalness? That i cannot say

smallp-o-p committed:
Remove missed CODESET macro wrapper remove, drop consteval test

philnik777 commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,835 @@
+  static constexpr auto __unknown_idx_ = 1u;
+
+  unsigned short __encoding_idx_{1u};
+  char __name_[max_name_length + 1]{};
```
Do we want to switch these over? `std::string` has two representations because the first one caused the actual string part to be unaligned, and this does basically the same. While not guaranteeing alignment, it's much more likely (e.g. when heap-allocating one of these) to be aligned if we put the string first.

smallp-o-p committed:
Swap members

frederick-vs-ja committed:
Merge branch 'main' into 105373-implement-P1885R12-P2862R1

frederick-vs-ja committed:
Update release notes

frederick-vs-ja committed:
Merge branch 'main' into 105373-implement-P1885R12-P2862R1

philnik777 commented on libcxx/include/text_encoding:
```diff
@@ -0,0 +1,835 @@
+    constexpr static unsigned long long __num_aliases =
+        std::count(__aliases_string, __aliases_string + sizeof(__aliases_string), '\0') + 1;
+    unsigned short __table[__num_aliases];
+  };
```
This is incredibly expensive. We're spending almost 20ms here just to calculate the number of elements in the array (every time this is included, even if not actually used). Can we replace the `std::count` call instead and just check that the number is correct with a libc++-specific test?

philnik777 replied:
@smallp-o-p Could you take a look at this?

ldionne commented:
Green Dragon
--------------
Actually, I just synced up with @Teemperor offline and we think the Green Dragon issue is caused by an old (and officially unsupported) SDK version on a Green Dragon builder. He's going to gate the failing test on the SDK version to hush that for now. So I don't think it's necessary to move anything to the dylib.

Picolibc
----------
@omjavaid @ceseo @antmox Could we get the picolibc bot updated so that what we're testing reflects something recent?

Cross Windows/Arm
---------------------
@vvereschaka I think this is going to be challenging to figure out without access to this configuration. What I'd suggest is adding this to the test temporarily:

```
// FIXME: something meaningful
// XFAIL: buildhost=windows && target=armv7-unknown-linux-gnueabihf
```

Then this can be removed when someone with access to this configuration has had time to investigate.
